Index: backports-6.5/backport-include/linux/etherdevice.h
===================================================================
--- backports-6.5.orig/backport-include/linux/etherdevice.h
+++ backports-6.5/backport-include/linux/etherdevice.h
@@ -39,7 +39,7 @@ static inline void u64_to_ether_addr(u64
 }
 #endif /* LINUX_VERSION_IS_LESS(4,11,0) */
 
-#if LINUX_VERSION_IS_LESS(5,15,0)
+#if LINUX_VERSION_IS_LESS(5,4,251)
 /**
  * eth_hw_addr_set - Assign Ethernet address to a net_device
  * @dev: pointer to net_device structure
