diff --git a/src/ap/wpa_auth_ft.c b/src/ap/wpa_auth_ft.c
index 99cf64f..a8dea61 100644
--- a/src/ap/wpa_auth_ft.c
+++ b/src/ap/wpa_auth_ft.c
@@ -3818,7 +3818,7 @@ pmk_r1_derived:
 	wpa_ft_set_session_timeout(sm->wpa_auth, sm->addr, session_timeout);
 
 	buflen = 2 + sizeof(struct rsn_mdie) + 2 + sizeof(struct rsn_ftie) +
-		2 + FT_R1KH_ID_LEN + 200;
+		2 + FT_R1KH_ID_LEN + 200 + 2 + 1;
 #ifdef CONFIG_MTK_IEEE80211BE
 	if (sm->dot11MultiLinkActivated)
 		buflen += 12;
@@ -3848,6 +3848,15 @@ pmk_r1_derived:
 	pos += ret;
 
 #ifdef CONFIG_MTK_IEEE80211BE
+/*As per WFA 11R-MLO Testcase RSNXE expected to be present in FT-Auth response of AP if STA send RSNXE
+* in FT-Auth Request Frame which is not defined as per SPEC but added as workaround which need
+* to be checked further from WFA
+*/
+	ret = wpa_write_rsnxe(conf, pos, end - pos);
+	if (ret < 0)
+		goto fail;
+	pos += ret;
+
 	if (sm->dot11MultiLinkActivated) {
 		wpa_printf(MSG_DEBUG, "FT: Add ML IE in authentication response");
 		ret = wpa_write_mlie(conf, mladdr, NULL, 0, pos, end - pos, true);
